go/vt/sqlparser/sql.y

```diff
@@ -441,7 +441,11 @@ opt_partition_clause:
   }
 
 set_statement:
-  SET comment_opt update_list
+  SET comment_opt update_list ',' charset_or_character_set charset_value force_eof
```

Comment:
I don't think we should have a `force_eof` here since it seems to me it would allow anything after the character set, which is more lenient than I think we want this to be.

More generally I'm somewhat surprised this doesn't introduce grammar conflicts since the update_list is also a comma-separated set of expressions. As such, I don't off-hand see how the parser can figure out whether to continue processing the expressions as part of the update list as opposed to the character set case.

I suppose if it works, it works, but can you please confirm you don't get any shift-reduce or reduce-reduce conflicts when you build the grammar?

I do, however, think we should have a test case that shows the grammar won't parse this:
`SET foo=1, NAMES 'utf8', bar=1`

Reply:
I think there's a more generic way to fix this. Let me spin up a PR.

Reply:
❤️ @sougou